tools/tests/form-control/validation.test.js

```diff
@@ -0,0 +1,169 @@
+    <gameface-form-control>
+        <input id="role" name="role" value="some name" required />
+        <input minimum="3" maximum="5" name="username" value="Valid" type="text" id="username" />
+        <input min="10" max="30" value="15" name="password" type="text" id="password" />
+        <input type="text" value="missing name" id="no-name" />
+        <button class="form-element" id="submit" type="submit">Login</button>
+    </gameface-form-control>
+    `;
+
+    const el = document.createElement('div');
+    el.innerHTML = template;
+    el.className = 'form-validation-test-wrapper';
+    el.style.position = 'absolute';
+    el.style.top = '200px';
+    el.style.left = '200px';
+
+    let currentElement = document.querySelector('.form-validation-test-wrapper');
+
+    if (currentElement) {
+        currentElement.parentElement.removeChild(currentElement);
+    }
+
+    document.body.appendChild(el);
+
+    return new Promise(resolve => {
+        waitForStyles(resolve);
+    });
+}
+
+
+describe('Form validation', () => {
+    afterAll(() => {
+        const tooltips = document.querySelectorAll('gameface-tooltip');
+        for(let i = 0; i < tooltips.length; i++) {
+            tooltips[i].parentElement.removeChild(tooltips[i]);
+        }
+
+        let currentElement = document.querySelector('gameface-currentElement');
+
+        if (currentElement) {
+            currentElement.parentElement.removeChild(currentElement);
+        }
+    });
+
+    beforeEach(function (done) {
+        setupFormValidationTestPage().then(done);
+    });
+
+    it('Should show value missing error', async() => {
+        const input = document.querySelector('#role');
+        input.value = '';
+
+        click(document.querySelector('#submit'));
+
+        return createAsyncSpec(() => {
+            const tooltip = document.querySelector('#input-role-error-tooltip');
+            assert(tooltip.style.display !== 'none', 'Tooltip was not displayed!');
+
+            const expected = errorMessages.get('valueMissing');
+            const received = getTextContent(tooltip);;
+            tooltip.hide();
+            assert(received === expected, `The error message is not correct. Expected: ${expected}. Received: ${received}.`);
+        });
+    });
+
+    it('Should show value too long error', async() => {
+        const username = document.querySelector('#username');
+        username.value = 'This is too long';
+
+        click(document.querySelector('#submit'));
+
+        return createAsyncSpec(() => {
+            const tooltip = document.querySelector('#input-username-error-tooltip');
+            assert(tooltip.style.display !== 'none', 'Tooltip was not displayed!');
+
+            const expected = errorMessages.get('tooLong');
+            const received = getTextContent(tooltip);;
+            tooltip.hide();
+            assert(received === expected, `The error message is not correct. Expected: ${expected}. Received: ${received}.`);
+        });

```

Comment:
This is almost the same as the `return createAsyncSpec` from the previous and from the next function with the exception of the tooltip id and the error name.

Reply:
Yes. Most of the specs are very similar but I prefer to write them more in a more descriptive way (though too verbose) for better readability. It's hard to completely understand what a test does if it is generated dynamically. It might be more efficient code-length-wise, but it makes it harder to understand. Giving the fact that these are tests - I don't see a problem if the test is longer if it is more understandable.

Reply:
Yes but now you need to change in each test case the `;` symbol :D. What will happen if you need to rewrite some logic in this code block? Copy-paste again?
Also for me it will more descriptive if it was
```
await verifyTooltipErrorVisibility('tooltip-selector', 'tooltip error')
```

However you can leave it for now if you don't want to change this. If we add more tests with the same check then we will change it to method.